File: examples/integration/annotations/label_set.py
from bokeh.io import save
from bokeh.models import ColumnDataSource, LabelSet, LinearAxis, Plot, Range1d

source = ColumnDataSource(data=dict(text=['one', 'two', 'three'],
                                    x1=[1,4,7],
                                    x2=[60,240,420]))

# Have to specify x/y range as labels aren't included in the plot area solver
plot = Plot(plot_width=600, plot_height=600,
            x_range=Range1d(0, 10), y_range=Range1d(0, 10),
            toolbar_location=None)

label_set1 = LabelSet(x='x1', y=2, x_offset=25, y_offset=25,
                      text="text", source=source,
                      text_font_size='50px', text_color='red', text_alpha=0.9,
                      text_baseline='bottom', text_align='left',
                      background_fill_color='green', background_fill_alpha=0.2,
                      angle=15, angle_units='deg',
                      render_mode='canvas')

label_set2 = LabelSet(x='x2', y=4, x_units='screen', x_offset=25, y_offset=25,
                      text="text", source=source,
                      text_font_size='50px', text_color='red', text_alpha=0.9,
                      text_baseline='bottom', text_align='left',
                      background_fill_color='green', background_fill_alpha=0.2,
                      angle=15, angle_units='deg',
                      render_mode='css')

plot.add_layout(LinearAxis(), 'below')
plot.add_layout(LinearAxis(), 'left')

plot.add_layout(label_set1)
plot.add_layout(label_set2)

save(plot)

File: bokehjs/test/integration/annotations/label_set.ts
import {display, fig} from "#framework/layouts"

import {LabelSet, HTMLLabelSet} from "@bokehjs/models/annotations"
import {ColumnDataSource} from "@bokehjs/models/sources"
import type {Constructor} from "@bokehjs/core/class"

describe("LabelSet annotation", () => {

  function plot<T extends LabelSet | HTMLLabelSet>(LabelSetCls: Constructor<T>) {
    const plot = fig([300, 300], {x_range: [0, 10], y_range: [0, 10]})

    const source = new ColumnDataSource({
      data: {
        text: ["First label", "Second label\nspanning two lines", "Third label\nspanning\nthree lines", null, "Not shown"],
        x1: [1, 3, 7, 9, NaN],
        y1: [7, 5, 3, 1, NaN],
        x2: [30, 70, 150, 200, NaN],
      },
    })

    const label_set0 = new LabelSetCls({
      x: {field: "x1"}, y: {field: "y1"},
      x_offset: -10, y_offset: 25,
      angle: 15, angle_units: "deg",
      text: {field: "text"},
      source,
      text_font_size: "24px", text_color: "red", text_alpha: 0.9, text_baseline: "bottom", text_align: "left",
      background_fill_color: "green", background_fill_alpha: 0.8,
      border_line_color: "blue",
    })

    const label_set1 = new LabelSetCls({
      x: {field: "x2"}, y: 1,
      x_units: "screen", y_units: "data",
      x_offset: 0, y_offset: -5,
      angle: 0.25, angle_units: "turn",
      text: {field: "text"},
      source,
      text_font_size: "18px", text_color: "black", text_alpha: 0.9, text_baseline: "top", text_align: "left",
      background_fill_color: "orange", background_fill_alpha: 0.8,
      border_line_color: "red", border_line_width: 3,
    })

    plot.add_layout(label_set0)
    plot.add_layout(label_set1)

    return plot
  }

  it("should support basic positioning using canvas rendering", async () => {
    await display(plot(LabelSet))
  })

  it("should support basic positioning using CSS rendering", async () => {
    await display(plot(HTMLLabelSet))
  })
})
